File: app/src/main/java/com/cajetan/youtubeplus/utils/YouTubeData.java
package com.cajetan.youtubeplus.utils;

import android.Manifest;
import android.accounts.AccountManager;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.AsyncTask;
import android.support.annotation.NonNull;
import android.util.Log;
import android.widget.Toast;

import com.cajetan.youtubeplus.data.VideoData;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.googleapis.extensions.android.gms.auth.GooglePlayServicesAvailabilityIOException;
import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException;
import com.google.api.client.http.HttpTransport;
import com.google.api.client.json.JsonFactory;
import com.google.api.client.json.jackson2.JacksonFactory;
import com.google.api.client.util.ExponentialBackOff;
import com.google.api.services.youtube.YouTube;
import com.google.api.services.youtube.YouTubeScopes;
import com.google.api.services.youtube.model.SearchListResponse;
import com.google.api.services.youtube.model.SearchResult;
import com.google.api.services.youtube.model.Video;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import pub.devrel.easypermissions.AfterPermissionGranted;
import pub.devrel.easypermissions.EasyPermissions;

import static android.app.Activity.RESULT_OK;

public class YouTubeData implements EasyPermissions.PermissionCallbacks {

    private GoogleAccountCredential mCredential;

    private static final String TAG = YouTubeData.class.getSimpleName();

    private static final long SEARCH_PAGE_SIZE = 20;

    private static final int REQUEST_ACCOUNT_PICKER = 1000;
    private static final int REQUEST_AUTHORIZATION = 1001;
    private static final int REQUEST_GOOGLE_PLAY_SERVICES = 1002;
    private static final int REQUEST_PERMISSION_GET_ACCOUNTS = 1003;

    private static final String PREF_ACCOUNT_NAME = "accountName";
    private static final String[] SCOPES = {YouTubeScopes.YOUTUBE_READONLY};

    private String mVideoId = "";
    private String searchQuery = "";
    private String searchPageToken = null;
    private List<VideoData> mVideoData;

    private RequestType mRequestType;

    private Activity mActivity;

//    TODO: think about a singleton here

    public YouTubeData(Activity parentActivity) {
        mCredential = GoogleAccountCredential.usingOAuth2(
                parentActivity.getApplicationContext(), Arrays.asList(SCOPES))
                .setBackOff(new ExponentialBackOff());

        mActivity = parentActivity;
    }

    public void receiveVideoData(String videoId) {
        mVideoId = videoId;
        mRequestType = RequestType.DATA_REQUEST;

        getResultsFromApi();
    }

    public void receiveSearchResults(String search, String pageToken) {
        searchQuery = search;
        searchPageToken = pageToken;
        mRequestType = RequestType.SEARCH_REQUEST;

        getResultsFromApi();
    }

    public void receiveFavouritesResults(List<VideoData> videoData) {
        mVideoData = videoData;
        mRequestType = RequestType.FAVOURITES_REQUEST;

        getResultsFromApi();
    }

    private void getResultsFromApi() {
        if (!isGooglePlayServicesAvailable())
            acquireGooglePlayServices();
        else if (mCredential.getSelectedAccountName() == null)
            chooseAccount();
        else if (!isDeviceOnline())
            Toast.makeText(mActivity, "The device is not online.", Toast.LENGTH_SHORT).show();
        else {
            if (mRequestType == RequestType.DATA_REQUEST)
                new VideoDataTask(mCredential).execute(mVideoId);
            else if (mRequestType == RequestType.SEARCH_REQUEST)
                new VideoSearchTask(mCredential).execute(searchQuery, searchPageToken);
            else if (mRequestType == RequestType.FAVOURITES_REQUEST)
                new FavouritesTask(mCredential).execute(mVideoData.toArray(new VideoData[0]));
        }
    }

    private class VideoSearchTask extends AsyncTask<String, Void, List<Video>> {
        private com.google.api.services.youtube.YouTube mService;
        private Exception mLastError = null;
        private String mNextPageToken = null;
        private String mPreviousPageToken = null;

        VideoSearchTask(GoogleAccountCredential credential) {
            HttpTransport transport = AndroidHttp.newCompatibleTransport();
            JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();

            mService = new com.google.api.services.youtube.YouTube.Builder(
                    transport, jsonFactory, credential)
                    .setApplicationName("YouTube Plus")
                    .build();
        }

        @Override
        protected List<Video> doInBackground(String... args) {
            Log.d(TAG, "Receiving search results..");

            try {
                String videoId = args[0];
                String nextPageToken = args[1];

                YouTube.Search.List searchList = mService.search()
                        .list("id")
                        .setMaxResults(SEARCH_PAGE_SIZE)
                        .setQ(videoId)
                        .setType("");

                if (nextPageToken != null) {
                    Log.d(TAG, "Next page token: " + nextPageToken);
                    searchList.setPageToken(nextPageToken);
                }

                SearchListResponse response = searchList.execute();

                this.mNextPageToken = response.getNextPageToken();
                mPreviousPageToken = response.getPrevPageToken();

                StringBuilder finalId = new StringBuilder();
                for (SearchResult r : response.getItems()) {
                    String id = r.getId().getVideoId();

//                  TODO: Some stuff with this, if id turns out to be null
//                  (new Video()).setSnippet(searchResults.get(0).getSnippet().);
                    if (id == null)
                        continue;

                    finalId.append(id);
                    finalId.append(',');
                }
                finalId.setLength(finalId.length()-1);

                return mService.videos()
                        .list("snippet,contentDetails")
                        .setId(finalId.toString())
                        .execute()
                        .getItems();
            } catch (Exception e) {
                Log.e("YouTubeData", "Exception: " + e.toString());
                mLastError = e;
                cancel(true);
                return null;
            }
        }

        @Override
        protected void onPostExecute(List<Video> results) {
            Log.d(TAG, "Passing the results..");

            if (mActivity instanceof VideoSearchListener)
                ((VideoSearchListener) mActivity).onSearchResultsReceived(results,
                        mNextPageToken, mPreviousPageToken);
            else
                throw new UnsupportedOperationException("Activity must implement VideoSearchListener.");
        }

        @Override
        protected void onCancelled() {
            if (mLastError != null) {
                if (mLastError instanceof GooglePlayServicesAvailabilityIOException) {
                    Toast.makeText(mActivity, "GooglePlayServices not available.", Toast.LENGTH_LONG).show();
                } else if (mLastError instanceof UserRecoverableAuthIOException) {
                    mActivity.startActivityForResult(
                            ((UserRecoverableAuthIOException) mLastError).getIntent(),
                            YouTubeData.REQUEST_AUTHORIZATION);
                } else {
                    Log.e("YouTubeData", "Error occurred: " + mLastError.getMessage());
                }
            } else {
                Log.e("YouTubeData", "Request cancelled");
            }
        }
    }

    private class VideoDataTask extends AsyncTask<String, Void, Video> {
        private com.google.api.services.youtube.YouTube mService;
        private Exception mLastError = null;

        VideoDataTask(GoogleAccountCredential credential) {
            HttpTransport transport = AndroidHttp.newCompatibleTransport();
            JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();

            mService = new com.google.api.services.youtube.YouTube.Builder(
                    transport, jsonFactory, credential)
                    .setApplicationName("YouTube Plus")
                    .build();
        }

        @Override
        protected Video doInBackground(String... videoIds) {
            try {
                return mService.videos()
                        // TODO: some space for performance improvement here
                        .list("snippet, contentDetails")
                        .setId(videoIds[0])
                        .execute()
                        .getItems()
                        .get(0);
            } catch (Exception e) {
                Log.e("YouTubeData", "Exception: " + e.toString());
                mLastError = e;
                cancel(true);
                return null;
            }
        }

        @Override
        protected void onPostExecute(Video output) {
            if (mActivity instanceof VideoDataListener)
                ((VideoDataListener) mActivity).onVideoDataReceived(output);
            else
                throw new UnsupportedOperationException("Activity must implement VideoDataListener.");
        }

        @Override
        protected void onCancelled() {
            if (mLastError != null) {
                if (mLastError instanceof GooglePlayServicesAvailabilityIOException) {
                    Toast.makeText(mActivity, "GooglePlayServices not available.", Toast.LENGTH_LONG).show();
                } else if (mLastError instanceof UserRecoverableAuthIOException) {
                    mActivity.startActivityForResult(
                            ((UserRecoverableAuthIOException) mLastError).getIntent(),
                            YouTubeData.REQUEST_AUTHORIZATION);
                } else {
                    Log.e("YouTubeData", "Error occurred: " + mLastError.getMessage());
                }
            } else {
                Log.e("YouTubeData", "Request cancelled");
            }
        }
    }

    private class FavouritesTask extends AsyncTask<VideoData, Void, List<Video>> {
        private com.google.api.services.youtube.YouTube mService;
        private Exception mLastError = null;

        FavouritesTask(GoogleAccountCredential credential) {
            HttpTransport transport = AndroidHttp.newCompatibleTransport();
            JsonFactory jsonFactory = JacksonFactory.getDefaultInstance();

            mService = new com.google.api.services.youtube.YouTube.Builder(
                    transport, jsonFactory, credential)
                    .setApplicationName("YouTube Plus")
                    .build();
        }

        @Override
        protected List<Video> doInBackground(VideoData... args) {
            Log.d(TAG, "Receiving search results..");

            try {
                StringBuilder finalId = new StringBuilder();
                for (VideoData data : args) {
                    finalId.append(data.getVideoId());
                    finalId.append(',');
                }
                finalId.setLength(finalId.length()-1);

                return mService.videos()
                        .list("snippet, contentDetails")
                        .setId(finalId.toString())
                        .execute().getItems();
            } catch (Exception e) {
                Log.e("YouTubeData", "Exception: " + e.toString());
                mLastError = e;
                cancel(true);
                return null;
            }
        }

        @Override
        protected void onPostExecute(List<Video> results) {
            Log.d(TAG, "Passing the results..");

            if (mActivity instanceof FavouritesDataListener)
                ((FavouritesDataListener) mActivity).onFavouritesReceived(results);
            else
                throw new UnsupportedOperationException("Activity must implement FavouritesDataListener.");
        }

        @Override
        protected void onCancelled() {
            if (mLastError != null) {
                if (mLastError instanceof GooglePlayServicesAvailabilityIOException) {
                    Toast.makeText(mActivity, "GooglePlayServices not available.", Toast.LENGTH_LONG).show();
                } else if (mLastError instanceof UserRecoverableAuthIOException) {
                    mActivity.startActivityForResult(
                            ((UserRecoverableAuthIOException) mLastError).getIntent(),
                            YouTubeData.REQUEST_AUTHORIZATION);
                } else {
                    Log.e("YouTubeData", "Error occurred: " + mLastError.getMessage());
                }
            } else {
                Log.e("YouTubeData", "Request cancelled");
            }
        }
    }

    public void onParentActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQUEST_GOOGLE_PLAY_SERVICES:
                if (resultCode != RESULT_OK) {
                    Toast.makeText(mActivity,
                            "This app requires Google Play Services. Please install " +
                                    "Google Play Services on your device and relaunch this app.",
                            Toast.LENGTH_LONG).show();
                } else {
                    getResultsFromApi();
                }

                break;
            case REQUEST_ACCOUNT_PICKER:
                if (resultCode == RESULT_OK && data != null &&
                        data.getExtras() != null) {
                    String accountName =
                            data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME);
                    if (accountName != null) {
                        SharedPreferences settings =
                                mActivity.getPreferences(Context.MODE_PRIVATE);
                        SharedPreferences.Editor editor = settings.edit();
                        editor.putString(PREF_ACCOUNT_NAME, accountName);
                        editor.apply();
                        mCredential.setSelectedAccountName(accountName);
                        getResultsFromApi();
                    }
                }
                break;
            case REQUEST_AUTHORIZATION:
                if (resultCode == RESULT_OK)
                    getResultsFromApi();

                break;
        }
    }

    @AfterPermissionGranted(REQUEST_PERMISSION_GET_ACCOUNTS)
    private void chooseAccount() {
        if (EasyPermissions.hasPermissions(
                mActivity, Manifest.permission.GET_ACCOUNTS)) {

            String accountName = mActivity.getPreferences(Context.MODE_PRIVATE)
                    .getString(PREF_ACCOUNT_NAME, null);
            if (accountName != null) {
                mCredential.setSelectedAccountName(accountName);
                getResultsFromApi();
            } else {
                // Start a dialog from which the user can choose an account
                mActivity.startActivityForResult(
                        mCredential.newChooseAccountIntent(),
                        REQUEST_ACCOUNT_PICKER);
            }
        } else {
            // Request the GET_ACCOUNTS permission via a user dialog
            Log.d(TAG, "Requesting account permission");

            EasyPermissions.requestPermissions(
                    mActivity,
                    "This app needs to access your Google account (via Contacts).",
                    REQUEST_PERMISSION_GET_ACCOUNTS,
                    Manifest.permission.GET_ACCOUNTS);
        }
    }

    /**
     * Checks whether the device currently has a network connection.
     *
     * @return true if the device has a network connection, false otherwise.
     */
    private boolean isDeviceOnline() {
        ConnectivityManager connMgr =
                (ConnectivityManager) mActivity.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connMgr.getActiveNetworkInfo();
        return (networkInfo != null && networkInfo.isConnected());
    }

    /**
     * Check that Google Play services APK is installed and up to date.
     *
     * @return true if Google Play Services is available and up to
     * date on this device; false otherwise.
     */
    private boolean isGooglePlayServicesAvailable() {
        GoogleApiAvailability apiAvailability =
                GoogleApiAvailability.getInstance();
        final int connectionStatusCode =
                apiAvailability.isGooglePlayServicesAvailable(mActivity);
        return connectionStatusCode == ConnectionResult.SUCCESS;
    }

    /**
     * Attempt to resolve a missing, out-of-date, invalid or disabled Google
     * Play Services installation via a user dialog, if possible.
     */
    private void acquireGooglePlayServices() {
        GoogleApiAvailability apiAvailability =
                GoogleApiAvailability.getInstance();
        final int connectionStatusCode =
                apiAvailability.isGooglePlayServicesAvailable(mActivity);

        if (apiAvailability.isUserResolvableError(connectionStatusCode))
            Toast.makeText(mActivity, "GooglePlayServices not available.", Toast.LENGTH_LONG).show();
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) { }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) { }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        EasyPermissions.onRequestPermissionsResult(
                requestCode, permissions, grantResults, this);
    }

    private enum RequestType {
        DATA_REQUEST,
        SEARCH_REQUEST,
        FAVOURITES_REQUEST,
    }

    public interface VideoDataListener {
        void onVideoDataReceived(Video videoData);
    }

    public interface VideoSearchListener {
        void onSearchResultsReceived(List<Video> results,
                                     String nextPageToken, String previousPageToken);
    }

    public interface FavouritesDataListener {
        void onFavouritesReceived(List<Video> results);
    }
}

File: app/src/main/java/com/cajetan/youtubeplus/utils/YouTubeData.kt
package com.cajetan.youtubeplus.utils

import android.Manifest
import android.accounts.AccountManager
import android.app.Activity
import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.preference.Preference
import android.preference.PreferenceManager
import android.util.Log
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.cajetan.youtubeplus.data.PlaylistData
import com.cajetan.youtubeplus.data.VideoData
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.api.client.extensions.android.http.AndroidHttp
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.googleapis.extensions.android.gms.auth.GooglePlayServicesAvailabilityIOException
import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.api.client.util.ExponentialBackOff
import com.google.api.services.youtube.YouTube
import com.google.api.services.youtube.YouTubeScopes
import com.google.api.services.youtube.model.PlaylistItem
import com.google.api.services.youtube.model.Video
import com.google.api.services.youtube.model.VideoListResponse
import org.jetbrains.anko.doAsync
import org.jetbrains.anko.uiThread
import pub.devrel.easypermissions.AfterPermissionGranted
import pub.devrel.easypermissions.EasyPermissions
import java.lang.UnsupportedOperationException
import java.util.Collections.emptyList
import java.util.concurrent.CancellationException

class YouTubeData(parentActivity: Activity, fragment: Fragment? = null) :
        EasyPermissions.PermissionCallbacks {

    companion object {
        private val TAG = "YouTubeData"

        private const val SEARCH_PAGE_SIZE = 20
        private const val REQUEST_ACCOUNT_PICKER = 1000
        private const val REQUEST_PERMISSION_GET_ACCOUNTS = 1003
        private const val REQUEST_AUTHORIZATION = 1001
        private const val REQUEST_GOOGLE_PLAY_SERVICES = 1002

        private const val PREF_ACCOUNT_NAME = "accountName"
        private val SCOPES: Array<String> = arrayOf(YouTubeScopes.YOUTUBE_READONLY)
    }

    private val mActivity = parentActivity
    private val mFragment = fragment

    private val mCredential: GoogleAccountCredential = GoogleAccountCredential
            .usingOAuth2(parentActivity.applicationContext, SCOPES.toList())
            .setBackOff(ExponentialBackOff())

    private var accountInitialised: Boolean = false

    private var service: YouTube = YouTube.Builder(AndroidHttp.newCompatibleTransport(),
                JacksonFactory.getDefaultInstance(), mCredential)
                .setApplicationName("YouTube Plus")
                .build()

    private var mVideoId = ""
    private var mPlaylistId = ""
    private var searchQuery = ""
    private var pageToken = ""
    private var mChannelId = ""
    private lateinit var mVideoData: List<VideoData>
    private lateinit var mPlaylistsData: List<PlaylistData>
    private lateinit var mRequestType: RequestType

    private var favouritesCallback: ((List<Video>) -> List<Video>)? = null

    fun receiveVideoData(videoId: String) {
        mVideoId = videoId
        mRequestType = RequestType.DATA_REQUEST

        getResultsFromApi()
    }

    fun receiveSearchResults(search: String, pageToken: String) {
        searchQuery = search
        this.pageToken = pageToken
        mRequestType = RequestType.SEARCH_REQUEST

        getResultsFromApi()
    }

    fun receiveVideoListResults(videoData: List<VideoData>,
                                block: ((List<Video>) -> List<Video>)? = null) {
        mVideoData = videoData
        mRequestType = RequestType.VIDEO_LIST_REQUEST
        favouritesCallback = block

        getResultsFromApi()
    }

    fun receivePlaylistsLibraryResults(playlistsData: List<PlaylistData>) {
        mPlaylistsData = playlistsData
        mRequestType = RequestType.PLAYLISTS_LIBRARY_REQUEST

        getResultsFromApi()
    }

    fun receiveMostPopularResults(pageToken: String) {
        this.pageToken = pageToken
        mRequestType = RequestType.MOST_POPULAR_REQUEST

        getResultsFromApi()
    }

    fun receivePlaylistResults(playlistId: String, pageToken: String = "") {
        mPlaylistId = playlistId
        this.pageToken = pageToken
        mRequestType = RequestType.PLAYLIST_DATA_REQUEST

        getResultsFromApi()
    }

    fun receiveUploadPlaylistId(channelId: String) {
        mChannelId = channelId
        mRequestType = RequestType.UPLOAD_PLAYLIST_ID_REQUEST

        getResultsFromApi()
    }

    private fun getResultsFromApi() {
        when {
            !isGooglePlayServicesAvailable() -> acquireGooglePlayServices()
            mCredential.selectedAccountName == null && !accountInitialised -> chooseAccount()
            mCredential.selectedAccountName == null && accountInitialised -> { }
            !isDeviceOnline() -> Toast.makeText(mActivity, "The device is not online",
                    Toast.LENGTH_SHORT).show()
            else -> when (mRequestType) {
                RequestType.DATA_REQUEST -> videoDataTask(mVideoId)
                RequestType.SEARCH_REQUEST -> videoSearchTask(searchQuery, pageToken)
                RequestType.VIDEO_LIST_REQUEST -> videoListTask(mVideoData)
                RequestType.MOST_POPULAR_REQUEST -> mostPopularTask(pageToken)
                RequestType.PLAYLIST_DATA_REQUEST -> playlistDataTask(mPlaylistId, pageToken)
                RequestType.PLAYLISTS_LIBRARY_REQUEST -> playlistLibraryTask(mPlaylistsData)
                RequestType.UPLOAD_PLAYLIST_ID_REQUEST -> uploadPlaylistIdTask()
            }
        }
    }

    private fun videoSearchTask(videoId: String, pageToken: String) {
        doAsync {
            val result: ArrayList<FeedItem> = ArrayList()
            val nextPageToken: String
            val prevPageToken: String

            try {
                val searchList: YouTube.Search.List = service.search()
                        .list("id")
                        .setMaxResults(SEARCH_PAGE_SIZE.toLong())
                        .setQ(videoId)

                if (pageToken != "")
                    searchList.pageToken = pageToken

                val response = searchList.execute()

                nextPageToken = response.nextPageToken ?: ""
                prevPageToken = response.prevPageToken ?: ""

                // TODO: include channels and playlist in search results
                val finalVideoIds = StringBuilder()
                val finalPlaylistIds = StringBuilder()
                val finalChannelIds = StringBuilder()

                for (r in response.items) {
                    when (r.id.kind) {
                        "youtube#video" -> {
                            finalVideoIds.append(r.id.videoId)
                            finalVideoIds.append(',')
                        }
                        "youtube#playlist" -> {
                            finalPlaylistIds.append(r.id.playlistId)
                            finalPlaylistIds.append(',')
                        }
                        "youtube#channel" -> {
                            finalChannelIds.append(r.id.channelId)
                            finalChannelIds.append(',')
                        }
                    }
                }

                if (finalVideoIds.isNotEmpty())
                    finalVideoIds.setLength(finalVideoIds.length - 1)

                if (finalPlaylistIds.isNotEmpty())
                    finalPlaylistIds.setLength(finalPlaylistIds.length - 1)

                if (finalChannelIds.isNotEmpty())
                    finalChannelIds.setLength(finalChannelIds.length - 1)

                result.addAll(service.channels().list("snippet,statistics")
                        .setId(finalChannelIds.toString()).execute().items
                        .map { FeedItem(it.id, channel = it) })

                result.addAll(service.playlists().list("snippet,contentDetails")
                        .setId(finalPlaylistIds.toString()).execute().items
                        .map { FeedItem(it.id, playlist = it) })

                result.addAll(service.videos().list("snippet,contentDetails")
                        .setId(finalVideoIds.toString()).execute().items
                        .map { FeedItem(it.id, video = it) })

            } catch (e: java.lang.Exception) {
                onTaskCancelled(e)
                Log.d("YouTubeData", "Exception $e")
                throw CancellationException()
            }

            uiThread {
                val listener: VideoSearchListener? = when {
                    mActivity is VideoSearchListener -> mActivity
                    mFragment is VideoSearchListener -> mFragment
                    else -> null
                }

                listener?.onSearchResultsReceived(result.toList(), nextPageToken, prevPageToken)
                        ?: throw UnsupportedOperationException("Parent must implement VideoSearchListener")
            }
        }
    }

    // TODO: consider merging videoListTask with videoDataTask

    private fun videoDataTask(videoId: String) {
        doAsync {
            val result: Video
            try {
                result = service.videos().list("snippet,contentDetails")
                        .setId(videoId).execute().items[0]
            } catch (e: java.lang.Exception) {
                onTaskCancelled(e)
                throw CancellationException()
            }

            uiThread {
                val listener: VideoDataListener? = when {
                    mActivity is VideoDataListener -> mActivity
                    mFragment is VideoDataListener -> mFragment
                    else -> null
                }

                listener?.onVideoDataReceived(result)
                        ?: throw UnsupportedOperationException("Activity must implement VideoDataListener")
            }
        }
    }

    private fun videoListTask(videoData: List<VideoData>) {
        doAsync {
            val finalId = StringBuilder()
            for (data in videoData) {
                finalId.append(data.videoId)
                finalId.append(',')
            }

            val result: List<Video> = if (finalId.isEmpty()) {
                emptyList()
            } else {
                service.videos().list("snippet,contentDetails")
                        .setId(finalId.toString()).execute().items
            }

            uiThread {
                val listener: VideoListDataListener? = when {
                    mActivity is VideoListDataListener -> mActivity
                    mFragment is VideoListDataListener -> mFragment
                    else -> null
                }

                listener?.onVideoListReceived(result.toList(), favouritesCallback)
                        ?: throw UnsupportedOperationException("Activity must implement VideoListDataListener")
            }
        }
    }

    private fun playlistLibraryTask(playlists: List<PlaylistData>) {
        doAsync {
            val finalId = StringBuilder()
            for (playlist in playlists) {
                finalId.append(playlist.playlistId)
                finalId.append(',')
            }

            val result: List<FeedItem> = if (finalId.isEmpty()) {
                emptyList()
            } else {
                service.playlists().list("snippet,contentDetails")
                        .setId(finalId.toString()).execute().items
                        .map { FeedItem(it.id, playlist = it) }
            }

            uiThread {
                val listener: PlaylistLibraryListener? = when {
                    mActivity is PlaylistLibraryListener -> mActivity
                    mFragment is PlaylistLibraryListener -> mFragment
                    else -> null
                }

                listener?.onPlaylistsReceived(result.toList())
                        ?: throw UnsupportedOperationException("Activity must implement PlaylistLibraryListener")
            }
        }
    }

    private fun mostPopularTask(pageToken: String) {
        doAsync {
            val result: List<Video>
            val nextPageToken: String
            val prevPageToken: String

            Log.d("MainActivity", "Running the request")
            try {
                val searchList: YouTube.Videos.List = service.videos()
                        .list("snippet,contentDetails")
                        .setChart("mostPopular")
                        .setRegionCode("GB")

                if (pageToken != "")
                    searchList.pageToken = pageToken

                val response: VideoListResponse = searchList.execute()

                nextPageToken = response.nextPageToken ?: ""
                prevPageToken = response.prevPageToken ?: ""

                result = response.items
            } catch (e: java.lang.Exception) {
                onTaskCancelled(e)
                throw CancellationException()
            }
            Log.d("MainActivity", "Request complete")

            uiThread {
                val listener: MostPopularListener? = when {
                    mActivity is MostPopularListener-> mActivity
                    mFragment is MostPopularListener -> mFragment
                    else -> null
                }

                listener?.onMostPopularReceived(result, nextPageToken, prevPageToken)
                        ?: throw UnsupportedOperationException("Activity must implement MostPopularListener")
            }
        }
    }

    private fun playlistDataTask(playlistId: String, pageToken: String) {
        doAsync {
            val result: List<PlaylistItem>
            val nextPageToken: String
            val prevPageToken: String

            try {
                val list = service.playlistItems().list("snippet,contentDetails")
                        .setMaxResults(20).setPlaylistId(playlistId)


                if (pageToken.isNotEmpty())
                    list.pageToken = pageToken

                val response = list.execute()

                nextPageToken = response.nextPageToken ?: ""
                prevPageToken = response.prevPageToken ?: ""

                result = response.items
            } catch (e: java.lang.Exception) {
                onTaskCancelled(e)
                throw CancellationException()
            }

            uiThread {
                val listener: PlaylistDataListener? = when {
                    mActivity is PlaylistDataListener-> mActivity
                    mFragment is PlaylistDataListener -> mFragment
                    else -> null
                }

                listener?.onPlaylistDataReceived(result.toList(), nextPageToken, prevPageToken)
                        ?: throw UnsupportedOperationException("Activity must implement PlaylistDataListener")
            }
        }
    }

    // service.channels().list("").execute().items[0].contentDetails.relatedPlaylists.uploads

    private fun uploadPlaylistIdTask() {
        doAsync {
            val playlistId: String
            val channelTitle: String

            try {
                val responseItem = service.channels().list("snippet,contentDetails")
                        .setId(mChannelId).execute()
                        .items[0]

                playlistId = responseItem.contentDetails.relatedPlaylists.uploads
                channelTitle = responseItem.snippet.title
            } catch (e: java.lang.Exception) {
                onTaskCancelled(e)
                throw CancellationException()
            }

            uiThread {
                val listener: UploadPlaylistListener? = when {
                    mActivity is UploadPlaylistListener -> mActivity
                    mFragment is UploadPlaylistListener -> mFragment
                    else -> null
                }

                listener?.onUploadPlaylistIdReceived(playlistId, channelTitle)
                        ?: throw UnsupportedOperationException("Activity must implement UploadPlaylistListener")
            }
        }
    }

    private fun onTaskCancelled(error: Exception) {
        when (error) {
            is GooglePlayServicesAvailabilityIOException ->
                Toast.makeText(mActivity, "GooglePlayServices not available", Toast.LENGTH_LONG).show()
            is UserRecoverableAuthIOException ->
                mActivity.startActivityForResult(error.intent, YouTubeData.REQUEST_AUTHORIZATION)
            else -> Log.e(TAG, "Error occurred ${error.message}")
        }
    }

    fun onParentActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            REQUEST_GOOGLE_PLAY_SERVICES -> {
                if (resultCode != RESULT_OK) {
                    Toast.makeText(mActivity,
                            "This app requires Google Play Services. Please install " +
                                    "Google Play Services on your device and relaunch this app.",
                            Toast.LENGTH_LONG).show()
                } else {
                    getResultsFromApi()
                }
            }
            REQUEST_ACCOUNT_PICKER -> {
                if (resultCode == RESULT_OK && data?.extras != null) {
                    val accountName = data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME)
                    if (accountName != null) {
                        val settings = PreferenceManager.getDefaultSharedPreferences(mActivity)

                        val editor = settings.edit()
                        editor.putString(PREF_ACCOUNT_NAME, accountName)
                        editor.apply()

                        mCredential.selectedAccountName = accountName
                        getResultsFromApi()
                    }
                }
            }
            REQUEST_AUTHORIZATION -> {
                if (resultCode == RESULT_OK)
                    getResultsFromApi()
            }
        }
    }

    @AfterPermissionGranted(REQUEST_PERMISSION_GET_ACCOUNTS)
    private fun chooseAccount() {
        if (EasyPermissions.hasPermissions(mActivity, Manifest.permission.GET_ACCOUNTS)) {
            val accountName = PreferenceManager.getDefaultSharedPreferences(mActivity)
                    .getString(PREF_ACCOUNT_NAME, null)

            if (accountName != null) {
                mCredential.selectedAccountName = accountName
                accountInitialised = true
                getResultsFromApi()
            } else {
                mActivity.startActivityForResult(mCredential.newChooseAccountIntent(),
                        REQUEST_ACCOUNT_PICKER)
            }
        } else {
            EasyPermissions.requestPermissions(mActivity,
                    "This app needs to access your Google account (via Contacts).",
                    REQUEST_PERMISSION_GET_ACCOUNTS, Manifest.permission.GET_ACCOUNTS)
        }
    }

    /**
     * Checks whether the device currently has a network connection.
     *
     * @return true if the device has a network connection, false otherwise.
     */
    private fun isDeviceOnline(): Boolean {
        val conMgr = mActivity.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val networkInfo = conMgr.activeNetworkInfo
        return networkInfo?.isConnected ?: false
    }

    /**
     * Check that Google Play services APK is installed and up to date.
     *
     * @return true if Google Play Services is available and up to
     * date on this device; false otherwise.
     */
    private fun isGooglePlayServicesAvailable(): Boolean {
        val apiAvailability = GoogleApiAvailability.getInstance()
        return apiAvailability.isGooglePlayServicesAvailable(mActivity) == ConnectionResult.SUCCESS
    }

    /**
     * Attempt to resolve a missing, out-of-date, invalid or disabled Google
     * Play Services installation via a user dialog, if possible.
     */
    private fun acquireGooglePlayServices() {
        val apiAvailability = GoogleApiAvailability.getInstance()
        val connectionStatusCode = apiAvailability.isGooglePlayServicesAvailable(mActivity)

        if (apiAvailability.isUserResolvableError(connectionStatusCode))
            Toast.makeText(mActivity, "GooglePlayServices not available", Toast.LENGTH_LONG).show()
    }

    override fun onPermissionsGranted(requestCode: Int, perms: MutableList<String>) {
        getResultsFromApi()
    }

    override fun onPermissionsDenied(requestCode: Int, perms: MutableList<String>) {}

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>,
                                            grantResults: IntArray) {
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions,
                grantResults, this)
    }

    private enum class RequestType {
        DATA_REQUEST,
        SEARCH_REQUEST,
        VIDEO_LIST_REQUEST,
        MOST_POPULAR_REQUEST,
        PLAYLIST_DATA_REQUEST,
        PLAYLISTS_LIBRARY_REQUEST,
        UPLOAD_PLAYLIST_ID_REQUEST
    }

    interface VideoDataListener {
        fun onVideoDataReceived(videoData: Video)
    }

    interface VideoSearchListener {
        fun onSearchResultsReceived(results: List<FeedItem>,
                                    nextPageToken: String, previousPageToken: String)
    }

    interface VideoListDataListener {
        fun onVideoListReceived(results: List<Video>,
                                block: ((List<Video>) -> List<Video>)? = null)
    }

    interface PlaylistLibraryListener {
        fun onPlaylistsReceived(results: List<FeedItem>)
    }

    interface PlaylistDataListener {
        fun onPlaylistDataReceived(results: List<PlaylistItem>,
                                   nextPageToken: String, previousPageToken: String)
    }

    interface MostPopularListener {
        fun onMostPopularReceived(results: List<Video>,
                                  nextPageToken: String, previousPageToken: String)
    }

    interface UploadPlaylistListener {
        fun onUploadPlaylistIdReceived(id: String, channelTitle: String)
    }
}
